File: qqbot-plugin-api/src/main/java/com/mieai/qqbot/plugin/api/EventService.java
package com.mieai.qqbot.plugin.api;

import java.util.Set;

/** Registers named handlers. An empty event-type set subscribes to every event. */
public interface EventService {
    EventSubscription subscribe(String handlerId, Set<String> eventTypes, PluginEventHandler handler);

    static EventService denied() {
        return (handlerId, eventTypes, handler) -> {
            throw new SecurityException("Plugin event subscription capability is not granted");
        };
    }
}

File: qqbot-plugin-api/src/main/kotlin/com/mieai/qqbot/plugin/api/EventService.kt
package com.mieai.qqbot.plugin.api


/** Registers named handlers. An empty event-type set subscribes to every event. */
fun interface EventService {
    fun subscribe(handlerId: String, eventTypes: Set<String>, handler: PluginEventHandler): EventSubscription

    companion object {
        fun denied(): EventService = EventService { _, _, _ ->
            throw SecurityException("Plugin event subscription capability is not granted")
        }
    }
}
